lib/liquid/tags/render.rb

```diff
@@ -0,0 +1,54 @@
+module Liquid
+  class Render < Tag
+    Syntax = /(#{QuotedString})#{QuotedFragment}*/o
+
+    attr_reader :template_name_expr, :attributes
+
+    def initialize(tag_name, markup, options)
+      super
+
+      raise SyntaxError.new(options[:locale].t("errors.syntax.render".freeze)) unless markup =~ Syntax
+
+      template_name = $1
+
+      @template_name_expr = Expression.parse(template_name)
+
+      @attributes = {}
+      markup.scan(TagAttributes) do |key, value|
+        @attributes[key] = Expression.parse(value)
+      end
+    end
+
+    def render_to_output_buffer(context, output)
+      # Though we evaluate this here we will only ever parse it as a string literal.
+      template_name = context.evaluate(@template_name_expr)
+      raise ArgumentError.new(options[:locale].t("errors.argument.include")) unless template_name
+
+      partial = PartialCache.load(
+        template_name,
+        context: context,
+        parse_context: parse_context
+      )
+
+      inner_context = context.new_isolated_subcontext
```

Comment:
@fw42 you asked what were some of the performance gains I did for carted. This was a big part I didn't mention either. So carted parsed templates on upload because it kept templates in memory. So a big thing I would do in this instance is shift everything out of render that allocates up in to the parse method that way in render everything was available to use. You could maybe consider this for frequently used templates in Shopify.

Reply:
Actually this isn't a good example as its specific to this request. In this case this is where a pool would come in. I'd have a pool of empty subcontexts already created. Then all this would do is grab one of the available and if there were none it would create one. Then at the end of this block there would be a method to reset the subcontext back to its empty state for reuse by the next request.